cql3/restrictions/statement_restrictions.cc

```diff
@@ -54,7 +54,9 @@ class statement_restrictions::initial_key_restrictions : public primary_key_rest
     bool _allow_filtering;
 public:
     initial_key_restrictions(bool allow_filtering)
-        : _allow_filtering(allow_filtering) {}
+        : _allow_filtering(allow_filtering) {
+        this->expression = conjunction{};
```

Comment:
Does this need to be a conjunction? I think `true` is enough, and it will get converted to a conjunction when another is merged into it (and we should optimize (true && x to x).

Reply:
I don't like the idea of make_conjunction automatically optimizing the expression, since that makes its result not reflect the input faithfully.  That makes debugging harder and will (once CQL admits a wider range of expressions) make error messages confusing.

An alternative is to make the caller of make_conjunction do the short-circuiting where desired.  But that's essentially what I'm doing here, just with less code.

Reply:
Let's start with the first part. Why create a conjunction initially? If you were coding the standard calculator parser example, you'd code 3 * (4+5) as (product (literal 4) (sum (literal 4) (literal 5))), but you wouldn't encode 5 as (product (literal 5)), you'd write it as (literal 5).

Reply:
For higher fidelity of analogy, let's assume that your calculator only has addition.  How would you represent these inputs: "5", "0", "5+0", ""?  Would any of them be equal to another?

Reply:
BTW, this particular expression is never merged to another, so I'm find with it being `true`.  There is, however, another `conjunction{}` in the code [here](https://github.com/scylladb/scylla/blob/0688f5c3f93c86d7efed195df20ed93e1e1833f0/cql3/restrictions/single_column_primary_key_restrictions.hh#L104), which I don't think should be replaced with `true`.

Reply:
"5": (literal 5)
"0": (literal 0)
"5+0": (+ (literal 5) (literal 0))

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
I notice you left out one important case I asked about... 😊

Reply:
BTW, I like that you're using a Lisp syntax, because in Lisp `()` is the same as `nil`. ;)

Reply:
"" doesn't translate to anything in a calculator, it's a syntax error.

Reply:
> "" doesn't translate to anything in a calculator, it's a syntax error.

Not in the old calculator code you inherited. The old code has a notion of "initial" expression, and you must find a way to represent that temporarily, until you rewrite it later.